File: GUI/backend/src/protocol.py
from __future__ import annotations

from enum import IntEnum, StrEnum
from typing import Literal

from pydantic import BaseModel, Field


class PortState(StrEnum):
    DISCONNECTED = "DISCONNECTED"
    ENUMERATING = "ENUMERATING"
    AUTHORIZED = "AUTHORIZED"
    QUARANTINED = "QUARANTINED"
    MANUAL_BLOCKED = "MANUAL_BLOCKED"


class ThreatCode(IntEnum):
    NO_FAULT = 0
    UNAUTHORIZED_HID = 1
    KEYSTROKE_BURST = 2
    COOLDOWN_BURST = 3
    DESCRIPTOR_OVERFLOW = 4
    CRC_TOKEN_FAULT = 5
    BIT_STUFF_ERROR = 6
    MANUAL_HOST_KILL = 7


THREAT_NAMES: dict[int, str] = {code.value: code.name for code in ThreatCode}


class PortStatus(BaseModel):
    port: int = Field(ge=0, le=3)
    state: PortState
    threat_code: int = Field(ge=0, le=7)
    vbus_mv: int = Field(ge=0)
    current_ma: int = Field(ge=0)
    vid: str
    pid: str
    dev_class: str
    int_class: str


class TelemetryFrame(BaseModel):
    type: Literal["telemetry"] = "telemetry"
    timestamp_ms: int
    ports: list[PortStatus]


class SecurityEvent(BaseModel):
    type: Literal["event"] = "event"
    timestamp_ms: int
    port: int = Field(ge=0, le=3)
    threat_code: int = Field(ge=0, le=7)
    threat_name: str
    state: PortState
    message: str


class SnapshotMessage(BaseModel):
    type: Literal["snapshot"] = "snapshot"
    mode: str
    events: list[SecurityEvent]
    telemetry: TelemetryFrame


class CommandType(StrEnum):
    ISOLATE_PORT = "isolate_port"
    RESTORE_PORT = "restore_port"
    CLEAR_FAULT = "clear_fault"


class CommandMessage(BaseModel):
    type: Literal["command"] = "command"
    cmd: CommandType
    port: int = Field(ge=0, le=3)

File: GUI/frontend/src/types/protocol.ts
export type PortState =
  | "DISCONNECTED"
  | "ENUMERATING"
  | "AUTHORIZED"
  | "QUARANTINED"
  | "MANUAL_BLOCKED";

export type CommandType = "isolate_port" | "restore_port" | "clear_fault";

export interface PortStatus {
  port: number;
  state: PortState;
  threat_code: number;
  vbus_mv: number;
  current_ma: number;
  vid: string;
  pid: string;
  dev_class: string;
  int_class: string;
}

export interface TelemetryFrame {
  type: "telemetry";
  timestamp_ms: number;
  ports: PortStatus[];
}

export interface SecurityEvent {
  type: "event";
  timestamp_ms: number;
  port: number;
  threat_code: number;
  threat_name: string;
  state: PortState;
  message: string;
}

export interface SnapshotMessage {
  type: "snapshot";
  mode: string;
  events: SecurityEvent[];
  telemetry: TelemetryFrame;
}

export interface CommandMessage {
  type: "command";
  cmd: CommandType;
  port: number;
}

export const THREAT_NAMES: Record<number, string> = {
  0: "NO_FAULT",
  1: "UNAUTHORIZED_HID",
  2: "KEYSTROKE_BURST",
  3: "COOLDOWN_BURST",
  4: "DESCRIPTOR_OVERFLOW",
  5: "CRC_TOKEN_FAULT",
  6: "BIT_STUFF_ERROR",
  7: "MANUAL_HOST_KILL",
};

export type HubMessage = TelemetryFrame | SecurityEvent | SnapshotMessage;

export function isSnapshot(message: HubMessage): message is SnapshotMessage {
  return message.type === "snapshot";
}

export function isTelemetry(message: HubMessage): message is TelemetryFrame {
  return message.type === "telemetry";
}

export function isEvent(message: HubMessage): message is SecurityEvent {
  return message.type === "event";
}
